Os/test/ut/task/TaskRules.cpp

```diff
@@ -0,0 +1,249 @@
+}
+
+
+void Os::Test::Task::Tester::Join::action(
+        Os::Test::Task::Tester &state //!< The test state
+) {
+    printf("--> Rule: Join\n");
+    TestTaskInfo joiner_task;
+    const U32 random_index = STest::Pick::lowerUpper(0, state.m_tasks.size() - 1);
+
+    // Pop the middle
+    std::shared_ptr<TestTaskInfo> other_task = state.m_tasks[random_index];
+    state.m_tasks.erase(state.m_tasks.begin() + random_index);
+
+    // Set the other argument
+    joiner_task.m_other = other_task;
+
+    joiner_task.start(&TestTaskInfo::joining_task);
+
+    // Ensure it starts in the correct state for joining
+    ASSERT_EQ(other_task->stage(), TestTaskInfo::Lifecycle::MIDDLE);
+
+    // Poke the task into the MIDDLE state
+    joiner_task.signal();
+    wait_for_state_with_timeout(joiner_task, TestTaskInfo::MIDDLE, 100);
+
+    // Make the other task move from MIDDLE state to end state and wait for JOINED state
+    other_task->signal();
+    wait_for_state_with_timeout(joiner_task, TestTaskInfo::END, 100);
+    ASSERT_EQ(Os::Task::getNumTasks(), TestTaskInfo::s_task_count) << "Task count miss-match";
+}
+
+// ------------------------------------------------------------------------------------------------------
+// Rule:  CheckState
+//
+// ------------------------------------------------------------------------------------------------------
+
+Os::Test::Task::Tester::CheckState::CheckState() :
+        STest::Rule<Os::Test::Task::Tester>("CheckState") {
+}
+
+
+bool Os::Test::Task::Tester::CheckState::precondition(
+        const Os::Test::Task::Tester &state //!< The test state
+) {
+    return true;
+}
+
+void Os::Test::Task::Tester::CheckState::action(
+        Os::Test::Task::Tester &state //!< The test state
+) {
+    printf("--> Rule: Check State\n");
+    std::shared_ptr<TestTaskInfo> task;
+    const U32 random_index = STest::Pick::lowerUpper(0, state.m_tasks.size());
+
+    // Allow small chance to check new task is in NOT_STARTED state
+    if (random_index >= state.m_tasks.size()) {
+        task = std::make_shared<TestTaskInfo>();
+    } else {
+        task = state.m_tasks[random_index];
+    }
+    ASSERT_EQ(task->m_task.getState(), task->m_state) << "Task state progression not as expected.";
+    ASSERT_EQ(Os::Task::getNumTasks(), TestTaskInfo::s_task_count) << "Task count miss-match";
+}
+
+// ------------------------------------------------------------------------------------------------------
+// Rule:  Delay
+//
+// ------------------------------------------------------------------------------------------------------
+
+Os::Test::Task::Tester::Delay::Delay() :
+        STest::Rule<Os::Test::Task::Tester>("Delay") {
+}
+
+bool Os::Test::Task::Tester::Delay::precondition(
+        const Os::Test::Task::Tester &state //!< The test state
+) {
+    return true;
+}
+
+void Os::Test::Task::Tester::Delay::action(
+        Os::Test::Task::Tester &state //!< The test state
+) {
+    printf("--> Rule: Delay\n");
```

Comment:
I'm not sure we need these print statements. The STest framework has a way to print out the rule applications for the random scenarios.